src/main/java/org/jenkinsci/plugins/workflow/cps/CpsBodyExecution.java

```diff
@@ -172,29 +180,86 @@ private Continuable createContinuable(CpsThread currentThread, CpsCallableInvoca
+                        LinearBlockHoppingScanner scanner = new LinearBlockHoppingScanner();
+                        scanner.setup(t.head.get());
+                        for (FlowNode node : scanner) {
+                            if (node.getId().equals(startNodeId)) {
+                                // this head is inside this body execution
+                                StepExecution execution = t.getStep();
+                                if (execution != null) {
+                                    executions.add(execution);
+                                }
+                                break;
+                            }
+                        }
+                    }
+                    result.set(executions);
+                } catch (Exception x) {
+                    result.setException(x);
+                }
+            }
+            @Override public void onFailure(Throwable t) {
+                result.setException(t);
+            }
+        });
+        try {
+            return result.get(1, TimeUnit.MINUTES);
+        } catch (ExecutionException | InterruptedException | TimeoutException x) {
+            // TODO access to CpsThreadGroup.threads must be restricted to the CPS VM thread, but the API signature does not allow us to return a ListenableFuture or throw checked exceptions
+            throw new RuntimeException(x);
+        }
     }
 
     @Override
     public boolean cancel(final CauseOfInterruption... causes) {
         // 'stopped' and 'thread' are updated atomically
-        final CpsThread t;
+        CpsThread t;
         synchronized (this) {
             if (isDone())  return false;   // already complete
-            stopped = new FlowInterruptedException(Result.ABORTED, causes); // TODO: the fact that I'm hard-coding exception seems to indicate an abstraction leak. Come back and think about this.
+            // TODO should perhaps rather override cancel(Throwable) and make this overload just delegate to that one
+            stopped = new FlowInterruptedException(Result.ABORTED, causes);
             t = this.thread;
         }
 
         if (t!=null) {
             t.getExecution().runInCpsVmThread(new FutureCallback<CpsThreadGroup>() {
                 @Override
                 public void onSuccess(CpsThreadGroup g) {
-                    t.stop(stopped);
+                    // Similar to getCurrentExecutions but we want the raw CpsThread, not a StepExecution; cf. CpsFlowExecution.interrupt
+                    Map<FlowHead, CpsThread> m = new LinkedHashMap<>();
+                    for (CpsThread t : thread.group.threads.values()) {
+                        m.put(t.head, t);
+                    }
+                    for (CpsThread t : Iterators.reverse(ImmutableList.copyOf(m.values()))) {
+                        LinearBlockHoppingScanner scanner = new LinearBlockHoppingScanner();
```

Comment:
Yeah, it's a micro-optimization here.  Stuff that's on the hot path should do that though wherever possible.